ServerCore/Pages/Submissions/FreeformQueue.cshtml.cs

```diff
@@ -0,0 +1,130 @@
+        {
+            if (puzzleId != null)
+            {
+                Puzzle = await _context.Puzzles.Where(m => m.ID == puzzleId).FirstOrDefaultAsync();
+            }
+
+            // todo: think about a way to go back to old submissions and redo them (e.g. to fix a forgotten comment)
+            // todo: should admins be able to see anything they're not an author of?
+            var submissionQuery = from Puzzle puzzle in _context.Puzzles
+                                  join PuzzleAuthors author in _context.PuzzleAuthors on puzzle.ID equals author.PuzzleID
+                                  join Submission submission in _context.Submissions on puzzle.ID equals submission.PuzzleID
+                                  join PuzzleStatePerTeam pspt in _context.PuzzleStatePerTeam on new { PuzzleID = puzzle.ID, TeamID = submission.TeamID } equals new { PuzzleID = pspt.PuzzleID, TeamID = pspt.TeamID }
+                                  where puzzle.EventID == Event.ID &&
+                                  author.Author == LoggedInUser &&
+                                  puzzle.IsFreeform &&
+                                  pspt.UnlockedTime != null && pspt.SolvedTime == null &&
+                                  submission.FreeformAccepted == null
+                                  orderby submission.TimeSubmitted
+                                  select new SubmissionView() { PuzzleId = puzzle.ID, PuzzleName = puzzle.Name, TeamName = pspt.Team.Name, SubmissionText = submission.SubmissionText, SubmissionId = submission.ID };
+            if (puzzleId != null)
+            {
+                submissionQuery = submissionQuery.Where(submissionView => submissionView.PuzzleId == puzzleId);
+            }
+
+            Submissions = await submissionQuery.ToListAsync();
+        }
+
+        public async Task<IActionResult> OnPostAsync(int? puzzleId)
+        {
+            if (Result == FreeformResult.None)
+            {
+                return RedirectToPage();
+            }
+
+            bool accepted = (Result == FreeformResult.Accepted);
+
+            Submission submission;
+            using (var transaction = _context.Database.BeginTransaction())
+            {
+                // todo admin access?
+                submission = await (from Submission sub in _context.Submissions
+                                               join PuzzleAuthors author in _context.PuzzleAuthors on sub.PuzzleID equals author.PuzzleID
+                                               where sub.ID == SubmissionID &&
+                                               sub.FreeformAccepted == null &&
+                                               author.Author == LoggedInUser
+                                               select sub).FirstOrDefaultAsync();
+
+                if (submission != null)
+                {
+                    submission.FreeformAccepted = accepted;
+                    submission.FreeformResponse = FreeformResponse;
```

Comment:
If FreeformResponse is null, this should just be either "Accepted" or "Rejected". Better than blank.

Reply:
Good idea. Done.